### PR TITLE
refactor(velodyne): implement generic velodyne decoder

### PR DESCRIPTION
**Pull request by ike-kazu**
## PR Type
- Improvement
## Related Links
- Continuation of #138 

## Description
Previously, there was a separate Velodyne decoder for each sensor. However, since the sensors function very similarly, this leads to large amounts of repeated code that is unnecessary and difficult to maintain.
This PR implements a generic decoder to handle all Velodyne sensors.
### Design
`VelodyneDecoder<SensorT>`
Generic decoder using templating to get sensor-specific information
`VLP16`
Class holding VLP16 specific information
`VLS128`
Class holding VLS128 specific information
## Review Procedure
1. `colcon test --event-handlers console_cohesion+ --packages-above nebula_common`
2. `code review`
## Remarks
Current Progress:
| Task                                            | Done? | Notes                        |
| ----------------------------------------------- | ----- | ---------------------------- |
| Add VLP16 and VLS128 support to generic decoder |✅ |                              |
| Add VLP32 support to generic decoder	           |✅| Tests are broken, see below. |
| General refactor                                |       |                              |
| Validate on data live	                          |✅ |                              |
| Evaluate performance                            |✅|                              |
## Pre-Review Checklist for the PR Author
**PR Author should check the checkboxes below when creating the PR.**
- [x] Assign PR to reviewer
## Checklist for the PR Reviewer
**Reviewers should check the checkboxes below before approval.**
- [ ] Commits are properly organized and messages are according to the guideline
- [ ] (Optional) Unit tests have been written for new behavior
- [ ] PR title describes the changes
## Post-Review Checklist for the PR Author
**PR Author should check the checkboxes below before merging.**
- [ ] All open points are addressed and tracked via issues or tickets
## CI Checks
- **Build and test for PR**: Required to pass before the merge.

### CODE REVIEW

**Comment by ike-kazu** [comment]
@bgilby59

**Comment by ike-kazu** [comment]
Thank you @bgilby59 for refactoring velodyne drivers!
I tested this and report the result about it below. Please confirm it!

# Pointcloud Output
|Field Type||
|----|----|
|azimuth|screenshot_from_2024-05-01_17-34-49_720|
|channel|screenshot_from_2024-05-01_17-37-36_720|
|distance|screenshot_from_2024-05-01_17-40-35_720|
|elevation|screenshot_from_2024-05-01_17-40-54_720|
|intensity|screenshot_from_2024-05-01_17-41-18_720|
|return_type|screenshot_from_2024-05-01_17-41-53_720|
|timestamp|screenshot_from_2024-05-01_17-42-03_720|

🔴 Please check/fix: The pointcloud colors are weird in specific field types (see below). Do you have any idea what could be causing this?

|Field Type| Ideal | now |
|----|----|----|
|return type| all pointcloud colors are same | some point color is deferent|
|timestamp| pointcloud colors are changed to vertical and azimuth direction | all pointcloud colors are same|

# Decoder Performance
timing_comparison_720
🟢Performance is looking great!

**Comment by ike-kazu** [comment]
# Progress Update: Reason why all vlp tests are broken
It has 2 main reasons
- coordinates type is double. We need to put them in float.
- Getting Azimuth code has bag (see below)


| Correct output | Now output |
| ---- | ---- |
|  Screenshot from 2024-05-13 13-24-06 |  Screenshot from 2024-05-13 13-23-41 |

🔴 Pointcloud Shape is looking good. It has bag on just azimuth angle.

**Comment by mojomex** [comment]
Thank you @ike-kazu! The point cloud fields are looking good, and it's great the performance became better already.
As for the test cases, I advised @bgilby59 to set the fields to `double` to test if that makes it better but it seens you're right, we should set them back to `float`. See [0195ef0](https://github.com/tier4/nebula/commit/0195ef037f071cfb858c484a3c1264b00293be3c) for reference.

Once everything is `float` again, the tests for VLP16 and VLS128 should pass. For VLP32, there is a difference of 16 points but the code looks correct, so we might just regenerate the tests for that one.

**Comment by ike-kazu** [comment]
# Progress Update
Thank you @mojomex  for your advises! I tried to change all `double` field into `float`. Then, i can confirm that test is succeeded in vlp128, but it is failured in vlp16 and vlp32.
Furthermore,  i tried test and confirm the failures at vlp32 and vlp128 in this commit [0195ef0](https://github.com/tier4/nebula/commit/0195ef037f071cfb858c484a3c1264b00293be3c). So, I think that I can solve the vlp16 test failure with past commit.
So, I will follow the steps below.

1. fix the vlp16 decoder with seeing commits between current and [0195ef0](https://github.com/tier4/nebula/commit/0195ef037f071cfb858c484a3c1264b00293be3c).
2. fix the vlp32 decorder by myself.

**Comment by ike-kazu** [comment]
# Progress Update
I tried the test with changing `EXPECT_EQ` into `ASSERT_EQ` at [these parts](https://github.com/tier4/nebula/blob/main/nebula_tests/velodyne/velodyne_ros_decoder_test_vlp32.cpp#L270)  in each lidar test file to debug about pointclouds mount differences. 
I confirm that the number of pointclouds is different in each lidar. (See below)
```
# about vlp16 test
Expected equality of these values:
3:   pp1->points.size()
3:     Which is: 14768
3:   pp2->points.size()
3:     Which is: 14736
```

```
# about vlp32 test
Expected equality of these values:
5:   pp1->points.size()
5:     Which is: 26196
5:   pp2->points.size()
5:     Which is: 26216
```

**Comment by ike-kazu** [comment]
# Progress Update
I reported about pointcloud number difference error in test. It was solved in [
d5fcf91](https://github.com/tier4/nebula/pull/144/commits/d5fcf91f24276aeec9b79450e7c27e7d920e6453). It is caused by filtering code for correct azimuth angle which is needed to get correct azimuth. Thank you @mojomex for giving me big advised.

We have still a problem about pointcloud coordinate matching. It is caused by the calculating resolution between `double` and `float`.

**Comment by ike-kazu** [comment]
# Progress Update
I reported test failure about vlp16. It was solved in [7395065].(https://github.com/tier4/nebula/pull/144/commits/73950655490d778789f774f03649183c0b6d6ac8).

Now only vlp32 test failure is remained.

**Comment by ike-kazu** [comment]
@mojomex 
hello! I have one question about nebula refactoring.
i don't know why [this function](https://github.com/ike-kazu/nebula/blob/feat/add-generic-velodyne-decoder/nebula_decoders/include/nebula_decoders/nebula_decoders_velodyne/decoders/vlp_16.hpp#L31) is needed in each velodyne lidar. In this function, true azimuth angle whch can pass nebula_test is selected. I create for nebula_test. To pass this test, I should refactor codes to be not using calculated accurate azimuth angles and using before caluculated in only vlp32. But other velodyne lidar vlp16 and vls128 test are passed with using calculated accurate angles. i checked vlp16 and vlp32 velodyne user manual and confirm that calculating accurate angle is needed. I think that vlp32c test data is suspicious. i would like to ask your opinion.
thank you

**Review by mojomex — CHANGES_REQUESTED** [review]
Hi @ike-kazu, thank you for your great work!
I have reviewed the changes and found a few more points that still need a bit of tweaking (a few of them I missed in my initial review of #138). Could you please implement the requested changes?

Thank you very much!

File: src/transport_drivers
Comment (mojomex) [inline src/transport_drivers:1]: Please remove this as a git submodule

**Comment by ike-kazu** [comment]
@mojomex 
I finished to code with your advices, Thank you! Please review.

**Comment by mojomex** [comment]
Will be continued in #228.

**Pull request closed without merging**